### ISSUE TITLE
Using Frozen Parameters to Freeze Differing Parameters Based Upon Epoch

### ISSUE THREAD

**Issue by adam-hartshorne** [patrick-kidger/equinox#266]
I am trying to figure out if it is possible to efficiently use something like the frozen parameters approach described here https://docs.kidger.site/equinox/examples/frozen_layer/ , to setup up a training loop such that a different subset of parameters are frozen based upon the epoch 

e.g For simplicity's sake, let's say you have an MLP, and you want the weights of odd and even layers to alternate between being updated each epoch (this isn't my real use case).  As far as I can tell, if you use the filter_spec approach, the opt_state will have None values for the frozen parameters, so you will then run into issues when you apply the opposite filter. Am I missing something here?

**Comment by anh-tong** [comment]
Hi, 

I work on something similar. For your example, what I can think of is to create two optimizers. Each optimizer transforms and update differently with different learning rates. This can be done by customizing learning rate schedule. Then we need to filter even and odd layer parameters. Finally, we can use `optax.multi_transform` (see [docs](https://optax.readthedocs.io/en/latest/api.html#optax.multi_transform)) to select which optimizers are used.

Here is a small example
```python
class Model(eqx.Module):
    
    even_layer: eqx.Module
    odd_layer: eqx.Module

def create_lr_scheduler(lr: float, is_even: bool, num_steps_per_epoch: int):
    
    def lr_scheduler(num_steps):
        current_epoch = num_steps // num_steps_per_epoch
        # return learning rate either 0 or the given learning rate based on epoch 
    
    return lr_scheduler

even_layer_optim = optax.adam(learning_rate=create_lr_scheduler(lr, is_even=True, num_steps_per_epoch=1000))
odd_layer_optim = optax.adam(learnning_rate=create_lr_scheduler(lr, is_even=False, num_steps_per_epoch=1000))


# first make all parameters with `even` labels
even_labels = jax.tree_utils.tree_map(lambda _: "even")

def get_odd_layer(tree: Model):
    # return all the odd layers. might be more complicated in your model
    return tree.odd_layer
    
# replace some odd params with new label
final_labels = eqx.tree_at(
    where=get_odd_layer,                # find where is the odd layer
    pytree=even_labels,
    replace_fn= lambda _: "odd"         # and replace with new labels
)

# the final optimizer will select specific optimizers based on the label
optim = optax.multi_transform(
    transforms={
        "even": even_layer_optim,
        "odd": odd_layer_optim,
    },
    param_labels=final_labels
)
```

Another solution is to use `optax.chain` with `optax.masked` (see [this](https://optax.readthedocs.io/en/latest/api.html#masked-update)) where we can mask on/off some of the parameters. But I find `optax.multi_transform` is more convenient where we can define many labels to do transformation.

**Comment by adam-hartshorne** [comment]
Thank you for the advice. I will give them a try.

**Comment by adam-hartshorne** [comment]
Edit - Silly question, with this multi_transform approach, how do you setup the training loop?

e.g. Normally with Equinox "model" you use a setup something like this,

```
opt_init, opt_update = optax.adam(config.learning_rate)
opt_state = opt_init(eqx.filter(model, eqx.is_inexact_array))
...
loss, grads = loss_func(model, x, y)
updates, update_opt_state = opt_update(grads, opt_state)
update_model = eqx.apply_updates(model, updates)
```
with the following, I get a "TypeError: Shapes must be 1D sequences of concrete values of integer type, got (None, 64)." upon the call to opt_init(...)
```
opt_init, opt_update = optax.multi_transform(
    transforms={
        "even": even_layer_optim,
        "odd": odd_layer_optim,
    },
    param_labels=final_labels
)

opt_state = opt_init(eqx.filter(model, eqx.is_inexact_array))
```

**Comment by anh-tong** [comment]
Uhm, I don't know why it throws `TypeError` here. Would you mind provide some more details (e.g., code, error trace)?

As far as I know, there is no further setup, and the training loop is the same with normal Equinox. I elaborate more in the example code here

```python
import equinox as eqx
import optax
import jax
import jax.tree_util


class Model(eqx.Module):
    
    even_layer: eqx.Module
    odd_layer: eqx.Module

def create_lr_scheduler(lr: float, is_even: bool, num_steps_per_epoch: int):
    
    def lr_scheduler(num_steps):
        current_epoch = num_steps // num_steps_per_epoch
        # return learning rate either 0 or the given learning rate based on epoch 
        return lr  # a dummy return
        
    
    return lr_scheduler

lr = 1e-3
even_layer_optim = optax.adam(learning_rate=create_lr_scheduler(lr, is_even=True, num_steps_per_epoch=1000))
odd_layer_optim = optax.adam(learning_rate=create_lr_scheduler(lr, is_even=False, num_steps_per_epoch=1000))

# assume that you model is initiated
model = Model(odd_layer=eqx.nn.Linear(3, 3, key=jax.random.PRNGKey(0)),
              even_layer=eqx.nn.Linear(3, 1, key=jax.random.PRNGKey(1)))

# first make all parameters with `even` labels
even_labels = jax.tree_util.tree_map(lambda _: "even", model)

def get_odd_layer(tree: Model):
    # return all the odd layers. might be more complicated in your model
    return tree.odd_layer
    
# replace some odd params with new label
final_labels = eqx.tree_at(
    where=get_odd_layer,                # find where is the odd layer
    pytree=even_labels,
    replace_fn= lambda _: "odd"         # and replace with new labels
)

# the final optimizer will select specific optimizers based on the label
optim_init, optim_update = optax.multi_transform(
    transforms={
        "even": even_layer_optim,
        "odd": odd_layer_optim,
    },
    param_labels=final_labels
)

optim_init(eqx.filter(model, eqx.is_inexact_array))
```
Here, `optim_init` works fine.

**Comment by anh-tong** [comment]
> Edit : How is "is_even" tested? Obviously, the labels are set to odd or even, how does the optimiser know "is_even" is true?

You can print out the `final_labels` to check how Pytree look likes. Unlike freezing layer where the leaves of freezed nodes in Pytree are `None`, the leaves of `final_labels` will be assigned with either label `even` or label `odd`. I guess that `multi_transform` using the the keys of the dictionary to select optimizers

```python
transforms={
        "even": even_layer_optim,
        "odd": odd_layer_optim,}
```
and `final_labels` will provide the information of parameters that helps optimizers find the coresponding parameters.

**Comment by adam-hartshorne** [comment]
OK, so I believe I have found the issue, but no idea why it should be a problem. Any attempt to override the __ call __ method results in various errors depending upon exactly what is being called from it. Define exactly the same code from a named method, no errors. 

Below is an (silly) example. I wonder if this is an equinox problem or an optax problem?

```
class Model(eqx.Module):
    odd_mlp: eqx.Module
    even_mlp: eqx.Module

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.odd_mlp = eqx.nn.MLP(in_size=2, out_size=2, width_size=32, depth=3, key=jax.random.PRNGKey(0))
        self.even_mlp = eqx.nn.MLP(in_size=2, out_size=2, width_size=32, depth=3, key=jax.random.PRNGKey(1))

    def calc_mlp(self, x1):
        return vmap(self.odd_mlp)(x1)

    # INCLUDE THE CALL OVERRIDE, GET ERROR
    # def __call__(self, x1):
    #     return vmap(self.odd_mlp)(x1)

def create_lr_scheduler(lr: float, is_even: bool, num_steps_per_epoch: int):
    def lr_scheduler(num_steps):
        current_epoch = num_steps // num_steps_per_epoch
        # return learning rate either 0 or the given learning rate based on epoch
        return lr  # a dummy return

    return lr_scheduler

lr = 1e-3
even_mlp_optim = optax.adam(learning_rate=create_lr_scheduler(lr, is_even=True, num_steps_per_epoch=1000))
odd_mlp_optim = optax.adam(learning_rate=create_lr_scheduler(lr, is_even=False, num_steps_per_epoch=1000))


# assume that you model is initiated
model = Model()

# first make all parameters with `even` labels
even_labels = jax.tree_util.tree_map(lambda _: "even", model)


def get_odd_mlp(tree: Model):
    # return all the odd layers. might be more complicated in your model
    return tree.odd_mlp


# replace some odd params with new label
final_labels = eqx.tree_at(
    where=get_odd_mlp,  # find where is the odd layer
    pytree=even_labels,
    replace_fn=lambda _: "odd"  # and replace with new labels
)

# the final optimizer will select specific optimizers based on the label
optim_init, optim_update = optax.multi_transform(
    transforms={
        "even": even_mlp_optim,
        "odd": odd_mlp_optim,
    },
    param_labels=final_labels
)

optim_init(eqx.filter(model, eqx.is_inexact_array))
```

**Comment by anh-tong** [comment]
Oh, I'm sorry. I encountered the problem before but don't recall. 

The solution here is to use different method name instead of '__call__'. 

'optax.multi_transform' takes 'param_labels' with 'Union[Any, Callable[[Any],Any]]'. During validation, it check if our passing value 'final_labels' is callable. If the Equinox module implement '__call__', it recognize 'final_labels' as callable. However, what we intend is to pass a Pytree with labels.

**Comment by adam-hartshorne** [comment]
Thanks for all your help, I now appear to have things running.

One final question, I am not sure I follow this part of the code. Why would you ever be returning a LR of 0?

```
 def lr_scheduler(num_steps):
        current_epoch = num_steps // num_steps_per_epoch
        # return learning rate either 0 or the given learning rate based on epoch 
        return lr  # a dummy return
```

**Comment by patrick-kidger** [comment]
Overriding `__call__`: this is a known problem with Optax; the reason is as @anh-tong describes.

If you want to keep the `__call__` method than a workaround is simply to wrap your model into a length-1 list; see #193 for an example of this issue in isolation.

As for the original problem of using different frozen parameters: probably the easiest approach is to pass in a mask indicating which parameters you want to update at each step, e.g.:

```python
def loss(model, x, y):
    pred_y = jax.vmap(model)(x)
    return jnp.mean((y - pred_y)**2)

@eqx.filter_jit
def make_step(model, opt_state, gradient_mask, x, y):
    grads = eqx.filter_grad(loss)(model, x, y)
    assert jtu.tree_structure(grads) == jtu.tree_structure(gradient_mask)
    grads = jtu.tree_map(lambda g, b: jnp.where(b, g, jnp.zeros_like(g)), grads, gradient_mask)
    updates, opt_state = optim.update(grads, opt_state)
    model = eqx.apply_updates(model, updates)
    return model, opt_state
```
here `gradient_mask` should be a PyTree with the same structure as `model`, with a `True/False` leaf for every floating-point array indicating whether that array should be updated with a gradient, and with `None` for all leaves that aren't floating-point arrays.

**Comment by anh-tong** [comment]
> Why would you ever be returning a LR of 0?

A quick implement can be
```python
def create_lr_scheduler(lr: float, is_even: bool, num_steps_per_epoch: int):
    
    def lr_scheduler(num_steps):
        current_epoch = num_steps // num_steps_per_epoch
        if is_even:
            return lr if current_epoch % 2 == 0 else 0.
        else:
            return lr if current_epoch % 2 == 1 else 0.
        
    return lr_scheduler
```

> or if in the loss function, there are numerous losses, different weighting of those that are normally passed as arguments to the loss function

Possibly, I think you can create an additional parameters in your Equinox module to work with the weight of losses.

> here `gradient_mask` should be a PyTree with the same structure as `model`, with a `True/False` leaf for every floating-point array indicating whether that array should be updated with a gradient, and with `None` for all leaves that aren't floating-point arrays.

Thanks @patrick-kidger for further clarification. `gradient_mask` seems very elegant.

**Comment by adam-hartshorne** [comment]
The code above results in a ConcretizationTypeError because an abstract tracer value was encountered where a concrete value is expected (due to the use of if / else). 

I think the gradient_mask approach is perfect for my particular use case.

**Comment by anh-tong** [comment]
Oh right, there is `ConcretizationTypeError`. So then, `gradient_mask` is the right approach.

**Comment by till-m** [comment]
Hi @patrick-kidger,

I know that the approach that you gave in this issue is not what is used in the [current equinox docs](https://docs.kidger.site/equinox/examples/frozen_layer/), so maybe it's not something you recommend anymore. I was planning to use it the approach here, as it's a bit less effort to implement, but wouldn't giving spurious zero-grads to Adam decay the momentum when it shouldn't?

**Comment by johannahaffner** [comment]
Yes, that could happen. The moving average of gradients for your frozen set of parameters would decay, which means that a) the frozen parameters would not truly be frozen, they should still get updated based on their past momentum - this affects the steps in whatever epoch you wish to freeze these parameters, and b) once you unfreeze them again, you will need to "pick up momentum" again.

So while you can think of freezing-per-epoch as a step function with solvers that do not depend on past information, with a solver such as Adam you get a smoothed and delayed version of the freeze-unfreeze step function.

**Comment by neel04** [comment]
Going back to the OP's original problem in the thread, AIUI if you want to have step/epoch dependent freezing in equinox then going with the approach laid out in the docs currently, we construct a `filter_spec` as the corresponding binary mask for the `PyTree` and partition the model accordingly.

Thus, the leaves at the frozen parameters are `None` as expected - however the optimiser state is initialised w.r.t the whole model and as a result contains valid `Array`s in place of the `None` values at those places causing problems with any subsequence `tree.map` calls.

Even if we meddle with the optimizer state, this is not a fantastic solution as in future we'd have to repair all the `PyTree`s if we wanna freeze a different part of the model.

I think an elegant solution here is much needed. cc @patrick-kidger

**Comment by patrick-kidger** [comment]
Hmm, I think that's probably an optimization problem more than a JAX/Equinox problem. Just in the mathematical abstract: if an optimizer is allowed to have arbitrary state then there isn't really a defined way to substitute out the optimization problem its solving, whilst maintaining that state.

**Comment by neel04** [comment]
I agree, this is a problem that's more-so suited to be solved by `optax` or other optimiser libraries - however `optax`'s current solution doesn't allow real dynamic freezing (i.e it resets the optimizer state) which isn't a great solution.

Something more elegant is simply `lax.stop_gradient` since it marks the output as a constant and thus never suffers from the aforementioned issue. The problem however is a way to unify the "freeze" attribute with the model's `PyTree`. I don't think there's a particularly elegant solution here that I've been able to find.

The outline, if it might be helpful to you, is to have a wrapper module that simply wraps around the output of the module we'd want to freeze. This wrapper, gated by a `flag: bool`, would wrap the value returned by `__call__` with `lax.stop_gradient`. This works fine, but the problem is there's no clean way to *change* the state of the `flag` under JIT.

The naive solution is that one just threads the argument for the flag through the entire model and suffers the re-compile (alleviated by `compilation_cache`-s). The ideal solution however would be to somehow mutate the flag field in the `PyTree` and have it obey it. This wouldn't require a reset of the `optimizer_state` since it's treated as a constant, so the values are `0` instead of `None` and thus also doesn't require additional filtering or an extra argument threaded everywhere.

Do you think such a solution could work? I think it'd just look very ugly and would have to modify the internals of the lib.

**Comment by johannahaffner** [comment]
You can wrap your boolean in a JAX array, or create a PyTree `freeze` whose leaves are boolean arrays and which is a tree-prefix of your model so as to allow for tree-mapped `where` filtering. Then you can filter out updates to the parameters you want to keep frozen with `where(frozen, 0, updates)`.
Boolean arrays do not trigger recompilation. Making this a model attribute isn't needed, but if so desired then you can make I6 one and update it with `eqx.tree_at`.

You might find [paramax](https://github.com/danielward27/paramax) useful as well.

**Comment by neel04** [comment]
> Then you can filter out updates to the parameters you want to keep frozen with where(frozen, 0, updates)

Yep, that's exactly what I meant by using the badly named `filter_spec` method. However, the problem is the optimizer state is still unconstrained and it's not exactly an elegant solution, with a lot of potential footguns.

> You can wrap your boolean in a JAX array

Sure, one could make an `eqx.nn.inference_mode` style API which I think would work. I think that's great for affecting the state of all the flags though and doesn't provide finer-grained control. I think having an equinox utility inspired from that which is able to provide finer-grained control for freezing (perhaps even built-in to `eqx.Module`'s `__call__`?) would be the correct solution.

Equinox is a library that's used mostly in research after all - I think it makes sense to improve the developer experience, especially for such a common requirement as dynamic freezing :)

**Comment by johannahaffner** [comment]
> > Then you can filter out updates to the parameters you want to keep frozen with where(frozen, 0, updates)
> 
> Yep, that's exactly what I meant by using the badly named `filter_spec` method. However, the problem is the optimizer state is still unconstrained and it's not exactly an elegant solution, with a lot of potential footguns.

What I meant is to use a tree-mapped `where` on the leaves, such that array elements become zeros rather than `None`. This way you do get fine-grained control without any recompilation. You can set any leaf to `jnp.array(True)` or `jnp.array(False)` or even pass a patterned boolean mask to filter elements of a leaf, at any time.
I'm suggesting that this be done on the dynamic part of your model, after filtering. 

The price to pay for such flexibility is keeping around another PyTree of the same shape as your model, or a tree-prefix of your model.
